app/code/Magento/InventoryCatalog/Plugin/InventoryApi/AssignSourcesToStock/PreventAssignSourcesToDefaultStock.php

```diff
@@ -0,0 +1,54 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\InventoryCatalog\Plugin\InventoryApi\AssignSourcesToStock;
+
+use Magento\InventoryApi\Api\AssignSourcesToStockInterface;
+use Magento\InventoryCatalog\Api\DefaultSourceProviderInterface;
+use Magento\InventoryCatalog\Api\DefaultStockProviderInterface;
+use Magento\InventoryCatalog\Exception\InvalidSourceAssignmentException;
+
+class PreventAssignSourcesToDefaultStock
+{
+    /**
+     * @var DefaultSourceProviderInterface
+     */
+    private $defaultSourceProvider;
+
+    /**
+     * @var DefaultStockProviderInterface
+     */
+    private $defaultStockProvider;
+
+    public function __construct(
+        DefaultSourceProviderInterface $defaultSourceProvider,
+        DefaultStockProviderInterface $defaultStockProvider
+    ) {
+        $this->defaultStockProvider = $defaultStockProvider;
+        $this->defaultSourceProvider = $defaultSourceProvider;
+    }
+
+    /**
+     * @param AssignSourcesToStockInterface $subject
+     * @param array $sourceIds
+     * @param int $stockId
+     * @return array
+     * @throws InvalidSourceAssignmentException
+     */
+    public function beforeExecute(AssignSourcesToStockInterface $subject, array $sourceIds, int $stockId)
+    {
+        if ($this->defaultStockProvider->getId() !== $stockId) {
```

Comment:
Looks like we can merge this two conditions

Reply:
>The mess detector complains about unused $subject parameter in the before plugin but we can't > avoid that parameter so what is the official guideline to deal with this kind of test failures?
Need to introduce special annotation `@SuppressWarnings(PHPMD.UnusedFormalParameter)`

Example
https://github.com/magento-engcom/msi/blob/develop/app/code/Magento/InventoryCatalog/Plugin/CatalogInventory/UpdateSourceItemAtLegacyQtyCounterPlugin.php#L87